metric-publishers/emf-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/emf/EmfMetricPublisher.java

```diff
@@ -0,0 +1,551 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.metrics.publishers.emf;
+
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.metrics.MetricCategory;
+import software.amazon.awssdk.metrics.MetricCollection;
+import software.amazon.awssdk.metrics.MetricLevel;
+import software.amazon.awssdk.metrics.MetricPublisher;
+import software.amazon.awssdk.metrics.MetricRecord;
+import software.amazon.awssdk.metrics.SdkMetric;
+import software.amazon.awssdk.thirdparty.jackson.core.JsonFactory;
+import software.amazon.awssdk.thirdparty.jackson.core.JsonGenerator;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * A metric publisher implementation that converts metrics into CloudWatch Embedded Metric Format (EMF).
+ * EMF allows metrics to be published through CloudWatch Logs using a structured JSON format, which
+ * CloudWatch automatically extracts and processes into metrics.
+ *
+ * <p>This publisher is particularly well-suited for serverless environments like AWS Lambda and container
+ * environments like Amazon ECS that have built-in integration with CloudWatch Logs. Using EMF eliminates
+ * the need for separate metric publishing infrastructure as metrics are automatically extracted from
+ * log entries.</p>
+ *
+ * <p>The EMF publisher converts metric collections into JSON-formatted log entries that conform to the
+ * CloudWatch EMF specification. These logs are written to the "/aws/emf/metrics" log group by default.
+ * CloudWatch automatically processes these logs to generate corresponding metrics that can be used for
+ * monitoring and alerting.</p>
+ *
+ * <p>Example usage:</p>
+ * <pre>
+ * EmfMetricPublisher publisher = PublisherBuilder.dimensions(CoreMetric.SERVICE_ID)
```

Comment:
We should use `@snippet` for code sample for javadoc

https://github.com/aws/aws-sdk-java-v2/blob/master/docs/design/APIReference.md#style-guideline